miniscope_io/device_update.py

```diff
@@ -1,144 +1,105 @@
-        if module == "EWL":
-            assert value <= ewlMAX and value >= ewlMIN
-    except AssertionError as msg:
-        if module == "LED":
-            logger.exception("LED value need to be a integer within 0-100")
-        if module == "EWL":
-            logger.exception("EWL value need to be an integer within 0-255")
-        raise msg
-
-    if module == "LED":
-        deviceTag = ledDeviceTag << deviceTagPos
-    elif module == "EWL":
-        deviceTag = ewlDeviceTag << deviceTagPos
-
-    command = [0, 0]
-
-    command[0] = int(
-        Preamble[0] * 2**preamblePos + deviceTag + np.floor(value / (2**uartPayload))
-    ).to_bytes(1, "big")
-    command[1] = int(
-        Preamble[1] * 2**preamblePos + deviceTag + value % (2**uartPayload)
-    ).to_bytes(1, "big")
-
-    # set up serial port
-    try:
-        serial_port = serial.Serial(port=comport, baudrate=baudrate, timeout=5, stopbits=1)
+        serial_port = serial.Serial(port=command.port, baudrate=2400, timeout=5, stopbits=2)
     except Exception as e:
         logger.exception(e)
         raise e
     logger.info("Open serial port")
 
-    for uartCommand in command:
-        for _ in range(uartRepeat):
-            # read UART data until preamble and put into queue
-            serial_port.write(uartCommand)
-            time.sleep(uartTimeGap)
+    try:
+        id_command = (command.device_id + UpdateCommandDefinitions.id_header) & 0xFF
+        serial_port.write(id_command.to_bytes(1, "big"))
+        logger.debug(f"Command: {format(id_command, '08b')}; Device ID: {command.device_id}")
+        time.sleep(0.1)
+
+        target_command = (command.target.value + UpdateCommandDefinitions.target_header) & 0xFF
+        serial_port.write(target_command.to_bytes(1, "big"))
+        logger.debug(f"Command: {format(target_command, '08b')}; Target: {command.target.name}")
+        time.sleep(0.1)
+
+        value_LSB_command = (
+            (command.value & UpdateCommandDefinitions.LSB_value_mask)
+            + UpdateCommandDefinitions.LSB_header
+        ) & 0xFF
+        serial_port.write(value_LSB_command.to_bytes(1, "big"))
+        logger.debug(f"Command: {format(value_LSB_command, '08b')}; Value: {command.value} (LSB)")
+        time.sleep(0.1)
+
+        value_MSB_command = (
+            ((command.value & UpdateCommandDefinitions.MSB_value_mask) >> 6)
+            + UpdateCommandDefinitions.MSB_header
+        ) & 0xFF
+        serial_port.write(value_MSB_command.to_bytes(1, "big"))
+        logger.debug(f"Command: {format(value_MSB_command, '08b')}; Value: {command.value} (MSB)")
+        time.sleep(0.1)
+
+        serial_port.write(UpdateCommandDefinitions.reset_byte.to_bytes(1, "big"))
+
+    finally:
+        serial_port.close()
+        logger.info("Closed serial port")
+
+
+def find_ftdi_device() -> list:
+    """
+    Find FTDI devices connected to the computer.
+    """
+    FTDI_VENDOR_ID = 0x0403
+    FTDI_PRODUCT_ID = 0x6001
```

Comment:
Probably not unless FTDI is in trouble.

Reply:
cool cool. only think i would suggest is moving constants to module level as a matter of style consistency, but idrc since i am not sure when else they would be needed and we don't have a code organization system where it would be obvious for someone to go looking for FTDI constants.

Reply:
OK, I'd usually try to think about putting specific variables in specific functions to save the namespace (and end up putting everything in global scope lol), but I don't really care for this case.

Reply:
definitely agreed globally scoped variables bad, and it's unfortunate there are not true consts in python. allcaps naming is not much of a protection ;)